pkg/kubelet/kubelet.go

```diff
@@ -2223,6 +2223,11 @@ func (kl *Kubelet) ListenAndServePodResources() {
 
 // Delete the eligible dead container instances in a pod. Depending on the configuration, the latest dead containers may be kept around.
 func (kl *Kubelet) cleanUpContainersInPod(podID types.UID, exitedContainerID string) {
+	// Handles container resource immediately.
+	if err := kl.containerManager.GetInternalContainerLifecycle().PostStopContainer(exitedContainerID); err != nil {
```

Comment:
To be sure I understand, running `PostStopContainer` immediately instructs the kubelet to immediately remove the container from the cpu manager... this would still happen later, but it would be part of a separate garbage collection process?

I have a couple questions:

1. Is there the risk the `PostStopContainer` will be updated to include actions that we wouldn't want to immediately run when we run `cleanUpContainersInPod`?
2. Is there a risk of conflict between the call to `PostStopContainer` and the garbage collection process we previously used? Or will the previous garbage collection be a no-op?

Reply:
`PostStopContainer` could be called at [`removeContainer`](https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/kuberuntime/kuberuntime_container.go#L804) and [`computePodActions`](https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/kuberuntime/kuberuntime_manager.go#L534), and both could happen to a same container, so I think it might not matter to call `PostStopContainer` twice at current situation.

> Is there the risk the PostStopContainer will be updated to include actions that we wouldn't want to immediately run when we run cleanUpContainersInPod?

I checked the code, and think it does not matter to run `PostStopContainer` immediately if the behavior  in `PostStopContainer` is idempotent.

> Is there a risk of conflict between the call to PostStopContainer and the garbage collection process we previously used? Or will the previous garbage collection be a no-op?

Yes, the `PostStopContainer` in the previous garbage collection will be a no-op.

>

Reply:
So, PostStopContainer is required because the CPU manager only removes resources for a container in PostStopContainer:

https://github.com/kubernetes/kubernetes/blob/f794c824b1e6e68b302d94f42f60af4759e18c6d/pkg/kubelet/cm/internal_container_lifecycle.go#L53-L55

But PostStopContainer is already called from the manager's removeContainer():

https://github.com/kubernetes/kubernetes/blob/f794c824b1e6e68b302d94f42f60af4759e18c6d/pkg/kubelet/kuberuntime/kuberuntime_container.go#L804

Furthermore, PostStopContainer is called from the main kubelet pod status control loop:

https://github.com/kubernetes/kubernetes/blob/f794c824b1e6e68b302d94f42f60af4759e18c6d/pkg/kubelet/kuberuntime/kuberuntime_manager.go#L537-L542

And that runs whenever the container state is not running.

So I'm wondering why this change is actually necessary. For some reason, the manager's removeContainer() isn't being called for this situation?

Reply:
@hex108 
There was no log in the following form for your kubelet:

"internal container post-stop lifecycle hook failed"

Reply:
@tedyu It never failed, it was just not called in time, so the CPU had not been released when the new Pod came.

Reply:
It seems Kubelet#syncPod might have run under this condition :
```
        if !runnable.Admit || pod.DeletionTimestamp != nil || apiPodStatus.Phase == v1.PodFailed {
```
which ends up with:
```
               return syncErr
```

Reply:
https://github.com/kubernetes/kubernetes/issues/79159#issuecomment-503482454 explained it. If not clear, I will add more explanation.

Reply:
Reading through this comment thread, it sounds a little to me like you're hitting a race condition, and by moving this functional call up a little you're now winning the race, but haven't actually addressed the problem. Is that the case?

Reply:
Finally had a chance to go through all this code. Turns out that the source of the race condition is the asynchronous pod container deletor. The only place that calls `Kubelet.cleanUpContainersInPod()` is `Kubelet.computePodActions` which fires `cleanUpContainersInPod()` when a pod lifecycle event is received.

`cleanUpContainersInPod` calls the pod container deletor's `deleteContainersInPod()` method which for each container in the pod ends up calling the kuberuntime_container's `DeleteContainer` method, which in turn calls the internal lifecycle manager's `PostStopContainer` method. The problem is that the pod container deletor is basically a work queue that is asynchronously processing requests to delete containers in a pod. Because of this, the DeleteContainer method ends up calling PostStopContainer asynchronously after `deleteContainersInPod` has returned successfully in `cleanUpContainersInPod`.

The solution to this race condition is to execute the code that we expect to be executed in a blocking fashion (the internal lifecycle manager's `PostStopContainer` callback) *before* calling the pod container deletor's `deleteContainersInPod` method. Which is what @hex108's code is doing here...

I would prefer that the code in `kuberuntime_container.removeContainer()` that calls `PostStopContainer` be *removed* instead of calling it twice, since leaving it in two different places is confusing.

Reply:
Actually, `Kubelet.HandlePodReconcile` also calls the asynchronous pod container deletor's `deleteContainers` method:

https://github.com/kubernetes/kubernetes/blob/e680ad7156f263a6d8129cc0117fda58602e50ad/pkg/kubelet/kubelet.go#L2135

and that doesn't specify a container ID and looks to rely on `DeleteContainer` eventually calling `PostStopContainer` :(

So I guess it's safest to leave PostStopContainer in the deleteContainer method..